Log parser time mean

peterfpeterson opened the pull request:
While walking through code I discovered that LogParser had a parallel implementation of calculating time average mean values. This merges it with the central version in TimeSeriesProperty.

**To test:**

The existing tests should all still pass.

Refs #34794.

*This does not require release notes* because it is a refactor to remove duplicate implementations.

<!-- Ensure the base of this PR is correct (e.g. release-next or main)
Finally, don't forget to add the appropriate labels, milestones, etc.!  -->

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

peterfpeterson committed:
Change TimeROI to be an optional pointer

peterfpeterson committed:
Use TimeSeriesProperty to find mean value

jmborr commented on Framework/Kernel/inc/MantidKernel/ITimeSeriesProperty.h:
```diff
@@ -43,13 +43,11 @@ class ITimeSeriesProperty {
   /// Return the time series's times as a vector<DateAndTime>
   virtual std::vector<Types::Core::DateAndTime> timesAsVector() const = 0;
-  /// Returns the calculated time weighted average value
-  virtual double timeAverageValue() const = 0;
```
you'll have to update the exposure of `timeAverageValue` to the python API in `Framework/PythonInterface/mantid/kernel/src/Exports/TimeSeriesProperty.cpp`

peterfpeterson replied:
I bet that is a couple of the broken tests right there. Thanks for the catch!

peterfpeterson committed:
Change the value of correct

peterfpeterson committed:
Use default value rather than overloading

peterfpeterson committed:
Fix overloaded export

peterfpeterson committed:
Remove unimplemented method

peterfpeterson committed:
Fix doxygen warning

peterfpeterson committed:
Fix doxygen warning

peterfpeterson committed:
Fix broken unit tests

Some of the test had their values of correct changed to reflect new
contribution of final time interval based on bugfix.

peterfpeterson committed:
Update test now that mean and time avg are equal

peterfpeterson committed:
Use all values in comparison

peterfpeterson committed:
Update the expected values for the tests

peterfpeterson committed:
Add release notes

peterfpeterson commented on Framework/Kernel/src/TimeSeriesProperty.cpp:
```diff
@@ -829,15 +819,18 @@ template <typename TYPE>
   if (realSize() == 1) {
     return static_cast<double>(m_values.front().value());
   }
+  if (size() == 1) {
```
It is a little more subtle than that. `realSize()` is the length of the vector and `size()` is the number of values with unique times. I haven't verified that it is enforced, but `size() <= realSize()` is an underlying assumption.

jmborr replied:
`size() <= realSize()` is my point. Condition `if (size() == 1)` makes `if (realSize() == 1)` superfluous.

peterfpeterson committed:
Remove redundate if case

peterfpeterson committed:
Update reference files

peterfpeterson committed:
"Fix" compiler warnings
